### PR TITLE
Johanna/5113 create new function app fhir test events

### PR DESCRIPTION
**Pull request by johanna-skylight**
# DEVOPS PULL REQUEST

## Related Issue

Implements #5113 and #5114

## Changes Proposed

The objective of this PR is to create a new function `FHIRTestEventReporter` within the `report_stream_batched_publisher` function app that will take care of reading and processing the FHIR multiplex test events pushed in the `fhir-data-publishing` queue. The code changes implemented take care of below items:
- Creates a new function within the function app `report_stream_batched_publisher`
- Modularizes the logic used by the function `QueueBatchedTestEventPublisher` so it can be reused by the new function
- Updates current unit testing to reflect the new logic split
- Implements the logic that takes care of reading the FHIR test events from `fhir-data-publishing` queue and process them into object type `ProcessedTestEvents`
- Implements new unit testing for the new logic 
- Add terraform changes to create the function and sets up env variables to refer to the queues `fhir-data-publishing` and `fhir-data-publishing-exceptions` used by the new function
- Updates the left-hook so the eslint checks, formatting and fixes trigger as pre-commit step

## Additional Information
- The logic to actually send the events to ReportStream will be done with ticket 5115 but you can see a rough implementation as part of this ticket but the code is commented out.
 
- You will see some code changes that are just formatting updates because of the left-hook triggering for updates in this folder moving forward.

- As part of this PR it has been decided that the reporting of FHIR test events will use its own queues to read and report failed parse test events but it will reuse the existing queue to log ReportStream response errors. 

## Testing
_This PR needs to be verified in a test environment (I am testing in dev4)_
- Verify that the existing logic to report test events works as expected
- Verify that the events published to `fhir-data-publishing` are correctly read and processed. These events will not be sent to the ReportStream yet so to validate this just check that the context logs are printed as expected in azure.

### CODE REVIEW

File: lefthook.yml
Comment (emmastephenson) [inline lefthook.yml:20]: Oooh thank you for spotting and fixing this gap! Should make developing on the function much better going forward 😄

File: ops/services/app_functions/report_stream_batched_publisher/functions/common/queueHandlers.test.ts
Comment (emmastephenson) [inline ops/services/app_functions/report_stream_batched_publisher/functions/common/queueHandlers.test.ts:33]: Github diffs are sadly no good at showing moved code - is it safe to assume all of this file was copied over from lib.test.ts?

File: ops/services/app_functions/report_stream_batched_publisher/functions/common/reportingHandlers.ts
Comment (emmastephenson) [inline ops/services/app_functions/report_stream_batched_publisher/functions/common/reportingHandlers.ts:18]: nit: I would name this something like `reportTestEventsToUniversalPipeline` to distinguish it from the existing ReportStream call, since they'll be used simultaneously for quite some time.

File: ops/services/app_functions/report_stream_batched_publisher/infra/main.tf
Comment (emmastephenson) [inline ops/services/app_functions/report_stream_batched_publisher/infra/main.tf:97]: nit: my preference would be to rename these queue names to start with FHIR so they're more easily distinguishable from the covid-only queues. I don't trust the Azure UI to make them easily distinguishable; it has a bad habit of cutting names off and not allowing you to expand them in different list views 🙃

**Review by emmastephenson — APPROVED** [review]
Overall this looks phenomenal, thank you Johanna!! Just a couple of small comments, but thanks for all your cleanup work here!

**Review by BobanL — DISMISSED** [review]
Changes look good! I'd agree with Emma on the nit's she pointed out.

File: ops/services/app_functions/report_stream_batched_publisher/functions/common/queueHandlers.test.ts
Comment (johanna-skylight) [reply ops/services/app_functions/report_stream_batched_publisher/functions/common/queueHandlers.test.ts:33]: yes, that is correct!. All the methods from lib.ts that were related to handling the queue were moved to the queueHandlers.ts file and all the unit tests from lib.test.ts that covered those methods was migrated to queueHandlers.test.ts. No coverage was added or removed in the migration :D.

**Review by emmastephenson — APPROVED** [review]
LGTM, thanks for addressing those nits!

**Comment by rin-skylight** [comment]
Currently undergoing DevOps review. Please do not merge just yet.

**Comment by rin-skylight** [comment]
The code looks good, from what I can see! As a question before approving: what plans, if any, are there to add alerting for function runs? Should that be included with this work?

**Comment by emmastephenson** [comment]
@rin-skylight that's a great question! We don't have specific tickets for this alerting yet, AFAIK. I would guess the eventual plan is to alert in a very similar manner to the existing function application (i.e., page if it fails to upload more than twice in a row) but some tweaks may be needed depending on how reliable the new ReportStream endpoint is. Since we're not yet sending data with this function app to ReportStream, my vote would be to hold off on adding any alerts until we've run the function with ReportStream for a week or two and assess what's needed at that time. I defer to @BobanL and @johanna-skylight on the final decision though!

**Review by emyl3 — COMMENTED** [review]
Was able to test multiplex being successfully added to the `fhir-data-publishing` queue. Is there a way you recommend testing/forcing an exception to see if it gets successfully added to the `fhir-data-publishing-exceptions` queue?? 🤔

**Comment by johanna-skylight** [comment]
> 

I don't think there was any plan to add alert for the function runs. Doing the research of the current queue implementation we did came to the conclusion that we need to add an alert to the `fhir-data-publishing-exceptions` queue that triggers when a test event gets pushed to it. I have created this ticket to track that https://github.com/CDCgov/prime-simplereport/issues/5183.

But if @emmastephenson or you have a specific alert in mind besides the one I mentioned please let me know!

**Comment by johanna-skylight** [comment]
> 

Manually push a message that will failed when parsing to JSON to the fhir queue!

**Comment by johanna-skylight** [comment]
@rin-skylight I have create a separate ticket to address the alerting https://github.com/CDCgov/prime-simplereport/issues/5187 Please let me know if this looks good for you!

CC: @emmastephenson let me know if the A/C is correct

**Review by rin-skylight — APPROVED** [review]
With the future work for the alerting piece captured, I have no further concerns. Clear from DevOps!

**Pull request merged**
